autoprecompiles/src/empirical_constraints.rs

```diff
@@ -0,0 +1,165 @@
+use std::collections::btree_map::Entry;
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+/// "Constraints" that were inferred from execution statistics. They hold empirically
+/// (most of the time), but are not guaranteed to hold in all cases.
+#[derive(Serialize, Deserialize, Clone, Default, Debug)]
+pub struct EmpiricalConstraints {
+    /// For each program counter, the range constraints for each column.
+    /// The range might not hold in 100% of cases.
+    pub column_ranges_by_pc: BTreeMap<u32, Vec<(u32, u32)>>,
+    /// For each basic block (identified by its starting PC), the equivalence classes of columns.
+    /// Each equivalence class is a list of (instruction index in block, column index).
+    pub equivalence_classes_by_block: BTreeMap<u64, BTreeSet<BTreeSet<(usize, usize)>>>,
```

Comment:
Writing this here before I forget: Thibaut explained to me how some of this works, and we had a discussion about it. One of the main points, which also caused some of my confusion in the `inteserct` function, is that the current collection of eq classes does not guarantee that a variable is in a single eq class (or not at all if it's a singleton). My initial thought was related to performance, as it doesn't sound great to always traverse the eq classes entirely to do the intersection (which may ofc also be negligible).